app/services/image_filename_service.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+class ImageFilenameService
+  attr_reader :document
+
+  def initialize(document)
+    @document = document
+  end
+
+  def call(suggested_name, mime_type)
+    normalised_name = normalise(suggested_name, mime_type)
+    ensure_unique(normalised_name)
+  end
+
+private
+
+  def images
+    @images ||= document.images
+  end
+
+  def normalise(name, type)
+    filename = ActiveStorage::Filename.new(name)
+    extension = Rack::Mime::MIME_TYPES.invert[type]
+    "#{filename.base.parameterize}#{extension}"
+  end
+
+  def ensure_unique(name)
+    bases = document.images.pluck(:filename)
+      .map(&ActiveStorage::Filename.method(:new))
+      .map(&:base)
+
+    filename = ActiveStorage::Filename.new(name)
+    return name unless bases.include?(filename.base)
+
+    suffix = unused_suffix(filename.base, bases)
+    "#{filename.base}-#{suffix}.#{filename.extension}"
+  end
+
+  def unused_suffix(suggested_base, bases)
+    suffixes = bases.map do |base|
+      match = base.match(/^#{suggested_base}-([0-9]+)$/)
+      match.to_a[1].to_i
+    end
+
+    ((1..(bases.count + 1)).to_a - suffixes).min
```

Comment:
Took me a mo to wrap my head round this method, clever!

Reply:
I'm happy to tweak this if you can think of something that makes it easier to understand :-).

Reply:
Hmm apart from generating a SecureRandom uuid for the suffix which isn't necessarily unique and makes horrendous file names, i can't think of a better approach then what you've implemented!

Reply:
I don't normally advocate comments in code, but in this case I think it is needed. Or, can you split this method up so that the child methods are named in a way that explains exactly what they are returning?

Reply:
As the method is only 4 lines long, there's not much scope for splitting it up. I do appreciate it's a little tricky to understand, though. Could you try and describe what's most confusing?

Reply:
Love the while loop! 

But @benthorner, to answer your question, the bits that I found confusing where:

`match.to_a[1].to_i`

and

`((1..(bases.count + 1)).to_a - suffixes).min`